### ISSUE TITLE
AppImage does not support HEIC/HEIF anymore

### ISSUE THREAD

**Issue by rdvdijk** [ImageMagick/ImageMagick#7475]
### ImageMagick version

7.1.1-24

### Operating system

Linux

### Operating system, version and so on

Ubuntu 20

### Description

Up until AppImage version 7.1.1-23, the AppImage release supported HEIC/HEIF.

Converting HEIC pictures using a version since 7.1.1-24 results in the following error message:

```
$ ./ImageMagick--gcc-x86_64-7.1.1-24.AppImage convert picture.heic picture.jpg
WARNING: The convert command is deprecated in IMv7, use "magick" instead of "convert" or "magick convert"

convert: Unsupported feature: Unsupported codec (4.3000) `/tmp/picture.heic' @ error/heic.c/IsHEIFSuccess/149.
convert: no images defined `picture.jpg' @ error/deprecate.c/ConvertImageCommand/3368.
```

We noticed that in 7.1.1-24 the AppImage the build environment was changed from Ubuntu 18 to Ubuntu 20. This might be the cause?

Looking at the build logs, it seems that `libheif` was actually found while building ImageMagick:

```
-------------------------------------------------------------
checking for libheif >= 1.4.0... yes
```
(from: https://github.com/ImageMagick/ImageMagick/actions/runs/9886317661/job/27305864652#step:7:310 )

### Steps to Reproduce

Try to convert any HEIC picture to JPEG. E.g:

```
$ ./ImageMagick--gcc-x86_64-7.1.1-24.AppImage convert picture.heic picture.jpg
```

**Comment by dlemstra** [comment]
The error message that you are seeing is from libheif. It is telling you that something in the file is not supported. Could also be that there was or is an issue with the latest release. And next time please include the file you are testing with so we are sure we are testing with the same files.

**Comment by Gillardo** [comment]
Any news on this, experiencing the same issue when trying to load in HEIC files taken on a iPhone 15 Pro Max

**Comment by dlemstra** [comment]
If you would have read my comment you would know not to ask for news on this. We don't know when we have time to work on this.

**Comment by gustavodiel** [comment]
Is libde265 installed and built as well? It's required for HEIC files

**Comment by liuatgit** [comment]
Tested many AppImage versions including latest 7.1.1-37 on AlmaLinux 9.4.  Not just HEIC/HEIF, it looks all formats require delegate (For example: AVIF) won't work after AppImage version 7.1.1-23.

**Comment by Sytten** [comment]
The AppImage is definitely not self-contained has it pretends @dlemstra. It is missing the plugins for `libheif`, I just spend a couple hours debugging this. The reason (I think) is that the version of libheif from the PPA is not listing it's dynamic dependencies:
```
@Sytten ➜ .../image2/squashfs-root/usr/lib $ ldd libheif.so.1 
        linux-vdso.so.1 (0x00007ffe811c5000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x000074ed32582000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x000074ed32566000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x000074ed32543000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x000074ed32361000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000074ed32212000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x000074ed321f7000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000074ed32003000)
        /lib64/ld-linux-x86-64.so.2 (0x000074ed32708000)
```

So the  AppImage creator won't fetch the needed libraries, but in reality if you built it yourself even without plugins with `cmake --preset=release-noplugins ..`, you see a different picture:
```
@Sytten ➜ .../codespaces-blank/libheif-1.19.2/build/libheif $ ldd libheif.so
        linux-vdso.so.1 (0x00007fffe0b8d000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007d1bac174000)
        libx265.so.179 => /lib/x86_64-linux-gnu/libx265.so.179 (0x00007d1baae92000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007d1bac151000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007d1baacb0000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007d1baab61000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007d1bac136000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007d1baa96f000)
        /lib64/ld-linux-x86-64.so.2 (0x00007d1bac186000)
        libnuma.so.1 => /lib/x86_64-linux-gnu/libnuma.so.1 (0x00007d1bac127000)
```

That makes sense if you look at https://github.com/strukturag/libheif?tab=readme-ov-file#supported-features, you need an x265 library to be able to use heic.

**Comment by fmw42** [comment]
I do not think that IM can build heif processing into the app image as some of the needed dependencies are not open source and require a different license.

**Comment by Sytten** [comment]
@fmw42 The heic support needs two things:
- Decoder: https://github.com/strukturag/vlc-libde265 which is LGPL, so can be built-into the AppImage with dynamic libs
- Encoder: https://github.com/videolan/x265 which is GPL, that is more problematic
The version 7.0.8 of ImageMagick used to STATIC link those two, that was a clear violation of GPL and LGPL.
But the decoding of HEIC can be shipped.

**Comment by Sytten** [comment]
Alright further digging confirms that the pre-built `libheif` looks for plugins ONLY in `/usr/lib/x86_64-linux-gnu/libheif/plugins` which makes it impossible to package the plugins into the `AppImage`. So we need to build `libheif` with at minimum the preset `release-noplugins` so the `.so` are linked directly. You would also need to disable the formats you dont need, see: https://github.com/strukturag/libheif/blob/master/CMakePresets.json#L121-L158

**Comment by fmw42** [comment]
I believe that libde265 is the issue

**Comment by Sytten** [comment]
I managed to make it work!

<details>
<summary>Build Script</summary>

```sh
#!/bin/bash
set -e
cd "$(dirname "$0")"

VERSION=1.19.2
SKIP_DEPS=false

# Parse arguments
for arg in "$@"; do
  if [[ ${arg} == "--skip-deps" ]]; then
    SKIP_DEPS=true
  fi
done

# Dependencies
if [[ ${SKIP_DEPS} == false ]]; then
  echo "[*] Installing dependencies"
  sudo apt update && sudo apt install libde265-dev -y
else
  echo "[*] Skipping dependency installation"
fi

# Source
if [[ ! -d "source" ]]; then
  echo "[*] Downloading source"
  wget -O source.tar.gz https://github.com/strukturag/libheif/releases/download/v"${VERSION}"/libheif-"${VERSION}".tar.gz
  tar -xvf source.tar.gz
  mv libheif-"${VERSION}" source
else
  echo "[*] Source already present, skipping download"
fi

# Build
echo "[*] Prepare build"
cp CMakePresets.json ./source/CMakePresets.json
rm -rf build
mkdir build && cd build
cmake --preset=basic ../source

echo "[*] Build"
make heif
```

</details>

<details>
<summary>CMake Preset</summary>

```json
{
  "version": 3,
  "cmakeMinimumRequired": {
    "major": 3,
    "minor": 0,
    "patch": 0
  },
  "configurePresets": [
    {
      "name": "basic",
      "displayName": "Release build without plugins",
      "description": "Release without plugins with minimal configuration for HEIC.",
      "cacheVariables": {
        "CMAKE_BUILD_TYPE": "Release",
        "BUILD_SHARED_LIBS": "ON",
        "BUILD_TESTING": "OFF",
        "ENABLE_EXPERIMENTAL_FEATURES": "OFF",
        "CMAKE_COMPILE_WARNING_AS_ERROR": "OFF",

        "ENABLE_PLUGIN_LOADING": "OFF",
        "WITH_AOM_DECODER": "OFF",
        "WITH_AOM_ENCODER": "OFF",
        "WITH_DAV1D": "OFF",
        "WITH_LIBDE265": "ON",
        "WITH_RAV1E": "OFF",
        "WITH_SvtEnc": "OFF",
        "WITH_X265": "OFF",
        "WITH_JPEG_DECODER": "OFF",
        "WITH_JPEG_ENCODER": "OFF",
        "WITH_UNCOMPRESSED_CODEC": "OFF",
        "WITH_KVAZAAR": "OFF",
        "WITH_OpenJPEG_DECODER": "OFF",
        "WITH_OpenJPEG_ENCODER": "OFF",
        "WITH_FFMPEG_DECODER": "OFF",
        "WITH_OpenH264_DECODER": "OFF",
        "WITH_UVG266": "OFF",
        "WITH_VVDEC": "OFF",
        "WITH_VVENC": "OFF",

        "WITH_REDUCED_VISIBILITY": "ON",
        "WITH_HEADER_COMPRESSION": "OFF",
        "WITH_LIBSHARPYUV": "OFF",
        "WITH_EXAMPLES": "OFF",
        "WITH_FUZZERS": "OFF"
      }
    }
  ]
}
```

</details>

It looks like the `CMakePresets.json` system [has a bug](https://github.com/strukturag/libheif/issues/1376) where `OFF` is ignored. Make sure that `libx265-dev` is NOT installed on the system otherwise it will be linked against it.

Once that is done, you get a nice so which links directly to `libde265.so.0`:
```
@Sytten ➜ .../functions/media-processor/libraries/libheif (ef-update-im) $ ldd build/libheif/libheif.so
        linux-vdso.so.1 (0x00007ffe740f2000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x000070b303f11000)
        libde265.so.0 => /lib/x86_64-linux-gnu/libde265.so.0 (0x000070b303ea0000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x000070b303bdd000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x000070b3039fb000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000070b3038ac000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x000070b303891000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000070b30369f000)
        /lib64/ld-linux-x86-64.so.2 (0x000070b303f23000)
```

**Comment by bradh** [comment]
I won't argue license interpretation, but if you want an encoder that is not x265, kvazaar could be an option.

**Comment by fmw42** [comment]
I think a developer will need to comment about both.

**Comment by jvdvegt** [comment]
This still seems to be an issue in the latest AppImage. @Sytten , did you push your changes somewhere so I can have a look? 

On Ubuntu 22.04, the issue is pretty simple to resolve. Add these two PPA's:
- https://launchpad.net/~strukturag/+archive/ubuntu/libheif
- https://launchpad.net/~strukturag/+archive/ubuntu/libde265

Install `libheif0` and `libde265-0`.

I could then use the latest AppImages (both nightly and 7.1.1-47)

**Comment by jvdvegt** [comment]
@Sytten yeah I saw your scripts, thanks. I just don't see what changes you made to libheif.
I wonder if merely building the AppImage on a system with an updated libheif (from the PPA I mention above) would result in an AppImage with libheif included. (Github builds them on a clean Ubuntu 20 image)

**Comment by maxnet** [comment]
>I wonder if merely building the AppImage on a system with an updated libheif (from the PPA I mention above) would result in an AppImage with libheif included.

As already mentioned by Sytten, the PPA has the problem that it contains plugins that are loaded dynamically at runtime.
The build process for an appimage does not know anything about plugins, and does not know it is supposed to bundle them in the appimage, so you end up with an appimage that is missing them. If it does work for you, that is because you happen to have the libheif plugins installed on your system, but that is not how appimage is supposed to work, it should work independent of what you have or have not on your system. 

Another problem with the PPA is that at the moment, the Ubuntu launchpad site that hosts them is unreachable (think due to a DDoS attack they are receiving, [status.canonical.com](https://status.canonical.com/) mentions: "Canonical’s web infrastructure is under a sustained, cross-border attack and we are working to address it").
Think it is better to build libheif from source as part of the github actions pipeline that currently builds ImageMagick, and will open up a PR to that extent.


>I just don't see what changes you made to libheif.

To get a flavor of libheif that is monolithic and does not put the functionality in separate plugins (that the appimage build process has problems locating), you do NOT need changes to the libheif code itself.
It only requires building it with different compile defines than were used by the PPA.
Choosing the build preset "release-noplugins" (which already exists, and I used in my PR), or creating a custom preset (like Sytten's custom "basic" preset, in his custom CMakepresets.json) will both achieve that.

**Comment by kmilos** [comment]
Btw, as an alternative to built-in codecs, it should be possible to 1) copy the plugins into the AppImage, and then 2) point to its specific location of the plugins by defining `LIBHEIF_PLUGIN_PATH` in the launcher.
